lib/tokens/coingecko/coingecko_test.go

```diff
@@ -15,47 +15,134 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+type testCase struct {
+	name         string
+	invalid      bool         // invalid response
+	empty        bool         // empty response
+	omitToken    tokens.Token // omit a requested token
+	renameToken  tokens.Token // rename a requested token
+	omitCurrency string       // omit a requested currency
+	zeros        bool         // include zero prices
+	negatives    bool         // include negative prices
+}
+
 func TestGetPrice(t *testing.T) {
 	t.Parallel()
 
-	// map token id -> currency -> price
+	tests := []testCase{
+		{name: "success"},
+		{name: "empty", empty: true},
+		{name: "omit eth", omitToken: tokens.ETH},
+		{name: "rename eth", renameToken: tokens.ETH},
+		{name: "omit omni", omitToken: tokens.OMNI},
+		{name: "rename omni", renameToken: tokens.OMNI},
+		{name: "omit usd", omitCurrency: "usd"},
+		{name: "zeros", zeros: true},
+		{name: "negatives", negatives: true},
+	}
+
+	shouldErr := func(t *testing.T, test testCase) bool {
+		t.Helper()
+		return (test.invalid ||
+			test.empty ||
+			test.omitToken != "" ||
+			test.renameToken != "" ||
+			test.omitCurrency != "" ||
+			test.zeros ||
+			test.negatives)
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+
+			server, servedPrices := makeTestServer(t, test)
+			defer server.Close()
+
+			c := coingecko.New(coingecko.WithHost(server.URL))
+			prices, err := c.Price(context.Background(), tokens.OMNI, tokens.ETH)
+
+			if shouldErr(t, test) {
+				require.Error(t, err)
+				return
+			}
+
+			require.NoError(t, err)
+			require.InEpsilon(t, prices[tokens.OMNI], servedPrices[tokens.OMNI.CoingeckoID()]["usd"], 0.01)
+			require.InEpsilon(t, prices[tokens.ETH], servedPrices[tokens.ETH.CoingeckoID()]["usd"], 0.01)
+		})
+	}
+}
+
+func makeTestServer(t *testing.T, test testCase) (*httptest.Server, map[string]map[string]float64) {
+	t.Helper()
+
 	// set during request handler
-	testPrices := make(map[string]map[string]float64)
+	servedPrices := make(map[string]map[string]float64)
 
-	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		require.Equal(t, "/api/v3/simple/price", r.URL.Path)
 
+		if test.invalid {
+			_, _ = w.Write([]byte("invalid json"))
+			return
+		}
+
 		q := r.URL.Query()
 		ids := strings.Split(q.Get("ids"), ",")
 		currencies := strings.Split(q.Get("vs_currencies"), ",")
 
 		resp := make(map[string]map[string]float64)
+
+		if test.empty {
+			bz, _ := json.Marshal(resp)
+			_, _ = w.Write(bz)
+
+			return
```

Comment:
Nit: if you happen to change the code before merging, consider moving this part up and put under handling `test.invalid` to avoid the unnecessary query parsing.